### ISSUE TITLE
JavaLauncher requires macOS 10.15

### ISSUE THREAD

**Issue by lfuelling** [perdian/macosappbundler-maven-plugin#7]
Hi, 
while testing the generated `.app` file on a friend's MacBook running the latest macOS 10.14, I got an error that the application requires macOS 10.15.

I assume the `JavaLauncher` which afaik is in `src/main/native` was compiled using 10.15.

I'd suggest using automated (Travis?) builds to compile the executable on multiple macOS versions and offering multiple versions of the app (where e.g. `1.3.0` is for the latest macOS, `1.3.0-hs` would be for high sierra, and so on) but this would probably cause errors in Maven repositories allowing `-SNAPSHOT` versions.

I also can understand if you don't care about this.

**Comment by perdian** [comment]
Should be pretty simple from I can see on the `clang` manpage (https://clang.llvm.org/docs/CommandGuide/clang.html):

```
Clang fully supports cross compilation as an inherent part of its design. Depending on how your version of Clang is configured, it may have support for a number of cross compilers, or may only support a native target.

[...]

-mmacosx-version-min=<version>

    When building for macOS, specify the minimum version supported by your application.
```

Will try to build this with a minimum version of 10.14 (Mohave) so let's see if this works.

**Comment by perdian** [comment]
Changed as part of https://github.com/perdian/macosappbundler-maven-plugin/commit/6b4f0355132802cfaea8a77547cc50b5c92a15e1

Released to Maven Central as `1.3.0`

**Comment by uwemock** [comment]
I just got a similar question from a user. He tells me that my application runs on MacOS 10.11.6 but not on 10.13.6. I would expect that if the bundle works on one version, it should work on later versions as well. And from this thread, I would expect it to work from version 10.14 and on.

My macosappbundler version is set to 1.3.0 in the pom.

Any ideas?

**Comment by perdian** [comment]
Version 1.3 defines macOS **10.14** as minimum version (see https://github.com/perdian/macosappbundler-maven-plugin/commit/6b4f0355132802cfaea8a77547cc50b5c92a15e1). 

So actually I wouldn't expect this to run on 10.11 at all. Strange macOS behaviour.

**Comment by uwemock** [comment]
I'm not familiar with MacOS at all...

So, is 10.14 considered a reasonable choice? As AdoptOpenJDK claims to support MacOS 10.10+, I would expect a bundler to support the same range of operating systems.

What is the current market share of 10.10 ... 10.14?

**Comment by lfuelling** [comment]
It's not a problem with the JDK, rather the executable that gets started when clicking the `.app` file (the stuff in [`src/main/native`](https://github.com/perdian/macosappbundler-maven-plugin/tree/master/src/main/native) I believe). 

As far as I understand it, the executable must either be compiled to support older versions or it can only support macOS versions upwards from what it was compiled on. 

To support 10.10, you'd need to compile the executable either on or for 10.10

**Comment by perdian** [comment]
I'm thinking about simply chaning the minimum version from 10.14 to 10.10 (analogous to https://github.com/perdian/macosappbundler-maven-plugin/commit/6b4f0355132802cfaea8a77547cc50b5c92a15e1).

Anyone think supporting older versions might be a problem? I can't really think of any issue but you never know ;-)

**Comment by uwemock** [comment]
I would recommend that, just to be in sync with the available OpenJDK versions.

However, I can't help with possible side effects - but I would give it a try.

**Comment by perdian** [comment]
> I would recommend that, just to be in sync with the available OpenJDK versions.

As @lfuelling wrote this has nothing to do with the JDK, it's just the native executable.

**Comment by uwemock** [comment]
I know.

But there is no point in having the (native executable in the) app run on 10.14+ when the JDK runs on 10.10+.

**Comment by lfuelling** [comment]
I also see no problems as long as the executable still runs on current stable versions. 

Unfortunately I have no running mac at the moment but I remember something about incompatibility with 32-bit apps in macOS Catalina.

**Comment by uwemock** [comment]
I have another issue.

Someone tells me about an error message "Cannot determine the location of JVM" in 10.12.6 and 10.14.6. JDK version in this case is Java 8u231.

Any ideas?

**Comment by perdian** [comment]
> Someone tells me about an error message "Cannot determine the location of JVM" in 10.12.6 and 10.14.6. JDK version in this case is Java 8u231.
> 
> Any ideas?

I could imagine that the JDK is installed at some exotic location, that the native launcher doesn't know about (or `/usr/libexec/java_home` isn't available at the expected location).

The detailed logics of where the launcher tries to find the JDK are here:
https://github.com/perdian/macosappbundler-maven-plugin/blob/master/src/main/native/helpers/jvm_resolver.m

**Comment by uwemock** [comment]
Funny detail: The version that I packaged with the appbundle-maven-plugin by federkasten works.

**Comment by perdian** [comment]
> Funny detail: The version that I packaged with the appbundle-maven-plugin by federkasten works.

Then check the logics from Federkasten and adjust the native launcher accordingly. Happy to merge a pull request when I get one.

**Comment by uwemock** [comment]
> 
> 
> > Funny detail: The version that I packaged with the appbundle-maven-plugin by federkasten works.
> 
> Then check the logics from Federkasten and adjust the native launcher accordingly. Happy to merge a pull request when I get one.

Honestly, I would love to. But I can't even tell what language you are using.

**Comment by lfuelling** [comment]
It's "Objective C" and since (afaik) you need a mac to code it I currently can't help with this either :(

**Comment by uwemock** [comment]
> It's "Objective C"

I expected that as I saw a few similarities to good old C. Unfortunately, most of it looks absolutely unfamiliar to me.

**Comment by uwemock** [comment]
> I could imagine that the JDK is installed at some exotic location, that the native launcher doesn't know about (or `/usr/libexec/java_home` isn't available at the expected location).

The user tells me everything is installed at the installer's default location

```
/Library/Application Support/Oracle/Java
/Library/Internet Plug-Ins/JavaAppletPlugin.plugin
/Library/PreferencePanes/JavaControPanel.prefPane
```

**Comment by uwemock** [comment]
I now have another user with the same problem. Same Java version (8u231), same error message. MacOS version is Catalina.

**Comment by uwemock** [comment]
...and another one on Mojave 10.14.6. Error message on 8u231, no problems on 9.

**Comment by perdian** [comment]
I understand that this isn't an ideal situation but as far as I'm concerned I'm not gonna spend any time in trying to fix this. It's pretty hard to debug without going back and installing the specific Java versions (looks like this is primarily an issue for Java 8 which is an LTS release but no longer in my focus).

So if anyone wants to pick this up and investigate, I'm happy to accept any pull request but I won't spend my time on this.

**Comment by uwemock** [comment]
I fully understand your point. I would be willing to provide all possible support. However, I can neither code in Objective C nor do I have any knowledge about MacOS. I could provide testing as I have some guys using my application.

The latest news is the output from the JavaLauncher executable:

```
~ $ /Applications/jes-mac.app/Contents/MacOS/JavaLauncher
2019-11-14 20:32:34.744 JavaLauncher[34069:745786] [INFO ] Looking up JVM directory for Java version: 8+
2019-11-14 20:32:34.808 JavaLauncher[34069:745786] [ERROR] Error status code returned from JVM home lookup: 2

```

**Comment by lfuelling** [comment]
@perdian I took a (quick) look at the code, would it be possible in Objective C to check the `JAVA_HOME` environment variable after the last log message in above output appears? 

eg. before [this throw here](https://github.com/perdian/macosappbundler-maven-plugin/blob/master/src/main/native/helpers/jvm_resolver.m#L50).

**Comment by uwemock** [comment]
Can we at least get the 10.10-compiled executable before long?
I see that I can get people to install Java 9, but system upgrades seem to be another story.

**Comment by perdian** [comment]
> Can we at least get the 10.10-compiled executable before long?
> I see that I can get people to install Java 9, but system upgrades seem to be another story.

I just released `1.3.2` to Maven central with macOS 10.10 as target version for the native binary (see 4276dab).

**Comment by uwemock** [comment]
Thanks. However, I read "1.3.1" in your commit, and Maven central shows me 1.3.0 as latest versoion.

**Comment by perdian** [comment]
> Thanks. However, I read "1.3.1" in your commit, and Maven central shows me 1.3.0 as latest versoion.

You're right, something seems to have gone wrong here. Will take another look at it.

**Comment by perdian** [comment]
Version `1.3.1` was what was actually released, and what is already available on Maven Central:

![grafik](https://user-images.githubusercontent.com/2276981/69351831-33a2b280-0c7c-11ea-93ea-eb7c380c1606.png)

Just to be safe I've re-release it as `1.3.2` (which is the identical source code base) which should be synced to Maven Central soon.

**Comment by uwemock** [comment]
Thanks. I was able to test this today, and I have found a possible drawback.
Usually, the user of the application got asked by the system if the program should really be launched. After having launched the application for the first time, the system didn't ask again on subsequent launches. With the new version, the system seems to ask each time the user launches the application.
Can this change in behavior be related to the changed compile flag?
